src/Diff.hs

```diff
@@ -0,0 +1,606 @@
+    underneath `/nix/store`, but this is the overwhelmingly common use case
+-}
+derivationName :: FilePath -> Text
+derivationName = Text.dropEnd 4 . Text.drop 44 . Text.pack
+
+-- | Group paths by their name
+groupByName :: Map FilePath a -> Map Text (Map FilePath a)
+groupByName m = Data.Map.fromList assocs
+  where
+    toAssoc key = (derivationName key, Data.Map.filterWithKey predicate m)
+      where
+        predicate key' _ = derivationName key == derivationName key'
+
+    assocs = fmap toAssoc (Data.Map.keys m)
+
+{-| Extract the name of a build product
+
+    Similar to `derivationName`, this assumes that the path name is:
+
+    > /nix/store/${32_CHARACTER_HASH}-${NAME}.drv
+-}
+buildProductName :: FilePath -> Text
+buildProductName = Text.drop 44 . Text.pack
+
+-- | Like `groupByName`, but for `Set`s
+groupSetsByName :: Set FilePath -> Map Text (Set FilePath)
+groupSetsByName s = Data.Map.fromList (fmap toAssoc (Data.Set.toList s))
+  where
+    toAssoc key = (buildProductName key, Data.Set.filter predicate s)
+      where
+        predicate key' = buildProductName key == buildProductName key'
+
+-- | Read a file as utf-8 encoded string, replacing non-utf-8 characters
+-- with the unicode replacement character.
+-- This is necessary since derivations (and nix source code!) can in principle
+-- contain arbitrary bytes, but `nix-derivation` can only parse from 'Text'.
+readFileUtf8Lenient :: FilePath -> IO Text
+readFileUtf8Lenient file =
+    Data.Text.Encoding.decodeUtf8With Data.Text.Encoding.Error.lenientDecode
+        <$> Data.ByteString.readFile file
+
+-- | Read and parse a derivation from a file
+readDerivation :: FilePath -> Diff (Derivation FilePath Text)
+readDerivation path = do
+    let string = path
+    text <- liftIO (readFileUtf8Lenient string)
+    case Data.Attoparsec.Text.parse Nix.Derivation.parseDerivation text of
+        Done _ derivation -> do
+            return derivation
+        _ -> do
+            fail ("Could not parse a derivation from this file: " ++ string)
+
+-- | Read and parse a derivation from a store path that can be a derivation
+-- (.drv) or a realized path, in which case the corresponding derivation is
+-- queried.
+readInput :: FilePath -> Diff (Derivation FilePath Text)
+readInput path =
+    if FilePath.isExtensionOf ".drv" path
+    then readDerivation path
+    else do
+        let string = path
+        result <- liftIO (Process.readProcess "nix-store" [ "--query", "--deriver", string ] [])
+        case String.lines result of
+            [] -> fail ("Could not obtain the derivation of " ++ string)
+            l : ls -> do
+                let drv_path = Data.List.NonEmpty.last (l :| ls)
+                readDerivation drv_path
+
+{-| Join two `Map`s on shared keys, discarding keys which are not present in
+    both `Map`s
+-}
+innerJoin :: Ord k => Map k a -> Map k b -> Map k (a, b)
+innerJoin = Data.Map.mergeWithKey both left right
```

Comment:
Sure, it can be replaced with `intersectionWith`, but this usage is safe and described in the same doc, as you refer on. This code was in the earlier version of nix-diff, I just didn't change it :)
So, will we change it?

Reply:
I don't really insist on changing it since this code existed previously